### ISSUE TITLE
Add low-rank / separable kernel primitive (LowRankKernel IR node)

### ISSUE THREAD

**Issue by jc-macdonald** [ACCIDDA/op_system#152]
## Summary

Add a first-class **low-rank / separable kernel** primitive to `op_system` so that contractions of the form

$$
y[i] = \sum_{j} K(i, j) \cdot x[j], \qquad K(i,j) = \sum_{c=1}^{R} w_c\, f_c(i)\, h_c(j)
$$

are evaluated as $R$ rank-1 contractions instead of materializing a dense $N \times N$ kernel matrix and performing an $O(N^2)$ matmul per RHS evaluation.

This is motivated by a concrete OOM and 80× CPU slowdown encountered in the COVID19_USA Round 19 continuum configs (63 age bins, separable contact kernel = 3 diagonal Gaussians + 1 symmetric parent-child pair, effective rank ≈ 6), but the primitive is generally useful: any kernel that is a finite sum of outer products (Gaussian mixtures, low-rank spectral approximations, HSGP basis evaluations) can use it.

## Motivation / impact (measured)

In the COVID19_USA `SMH_R19_op_system_continuum_hier_bspline_separable.yml` config:

- The force-of-infection is defined as an alias
  ```
  foi[age, loc] = r0[time] * r0_loc[loc] * gamma * chi_omicron
                * apply_along(contact_kernel[age, age:ap]
                              * apply_along(I/N, risk=r, vax=v, loc=l, kernel=sum),
                              age=ap, kernel=integrate)
  ```
- The 63×63 contact kernel is rank ≈ 6 (sum of separable Gaussian components), but `parametric.py` currently materializes it as a dense 63×63 matrix and op_system emits a full age×age contraction per cell.
- Combined with the alias-inlining issue (#153), this produces ~254 MB / draw of FOI intermediates and triggers an OOM at chunk size 8 on an 8.8 GB GPU (XLA demanded 12.6 GB and could not rematerialize below 12.59 GiB).

Per-draw memory if we factor the contraction:

| | Dense $N\times N$ kernel | Low-rank rank-$R$ |
|---|---|---|
| Kernel storage | $N^2$ = 16 KB (small) | $2 R N$ = ~3 KB |
| Per-cell intermediate | $O(N)$ × #cells | $O(R)$ × #cells |
| FOI intermediates (Round 19, $N=63$, #cells = 4 488) | **~254 MB** | **~20 MB** (92 % reduction) |

Wall-time on the same config: forward pass ~+5 %, VJP ~+20 % (smaller tape, fewer rematerializations).

### 1. New IR node

Add `LowRankKernel` (or `SeparableContraction`) to `op_system/_ir.py`:

```python
@dataclass(frozen=True)
class LowRankKernel(IRNode):
    contracted_axis: AxisRef             # e.g. age:ap
    output_axis: AxisRef                 # e.g. age
    factors_lhs: tuple[IRNode, ...]      # f_c(i), one per component, shape [R, N_out]
    factors_rhs: tuple[IRNode, ...]      # h_c(j), one per component, shape [R, N_in]
    weights: IRNode | None               # w_c, shape [R]; None => 1
    operand: IRNode                      # x[..., j, ...]  (with j == contracted_axis)
```

Semantics (NumPy reference):
```python
# operand has axis `j` of size N_in; factors have shape [R, N]
inner = jnp.einsum("rj,...j...->...r...", factors_rhs_stacked, operand)
if weights is not None:
    inner = inner * weights
out = jnp.einsum("ri,...r...->...i...", factors_lhs_stacked, inner)
```

### 2. Parser / spec-layer surface

Add a kernel mode in the parametric plugin spec so authors can declare separability without writing the contraction by hand:

```yaml
contact_kernel:
  type: low_rank_sum
  components:
    - { kind: gaussian, mu_lhs: 5,  sigma_lhs: 4,  mu_rhs: 5,  sigma_rhs: 4,  weight: w_pre }
    - { kind: gaussian, mu_lhs: 10, sigma_lhs: 3,  mu_rhs: 10, sigma_rhs: 3,  weight: w_sch }
    - { kind: gaussian, mu_lhs: 40, sigma_lhs: 12, mu_rhs: 40, sigma_rhs: 12, weight: w_ad  }
    - { kind: gaussian_pair_sym,
        mu_a: 30, sigma_a: 8, mu_b: 5, sigma_b: 4, weight: w_pc }
```

The parametric plugin returns `(factors_lhs, factors_rhs, weights)` instead of a materialized matrix, and the IR consumes them.

Alternatively (and more generally): expose a user-level operator
```
apply_low_rank_kernel(operand, kernel=foo, contracted=age:ap, output=age)
```
that the spec can call explicitly, with `foo` being any plugin output that has shape `(R, N)` factors.

### 3. Codegen

In `_vectorize.py` / `_ir_lower.py`, emit the two einsums above. Use `jnp.einsum` (or hand-rolled `jnp.sum(operand[..., None] * h, axis=j)`) so XLA fuses with surrounding pointwise ops.

### 4. Backward-compatibility

- Dense kernels keep working unchanged.
- The primitive only activates when the parametric plugin declares `type: low_rank_sum` (or when the user writes `apply_low_rank_kernel(...)`).
- For correctness validation, add a `materialize=True` debug flag that reconstructs $K = \sum_c w_c f_c h_c^\top$ and asserts it matches the dense path within fp32 tolerance.

## Acceptance criteria

- [ ] `LowRankKernel` IR node defined and round-trips through `_ir.py` / `_ir_expand.py` / `_vectorize.py`.
- [ ] Codegen produces two einsums (rank-$R$ inner reduction + rank-$R$ outer expansion) and no $N\times N$ intermediate.
- [ ] Numerical agreement with dense path on a unit test (Gaussian-mixture kernel, $N=63$, $R=6$) within `rtol=1e-5`.
- [ ] Benchmark: on the COVID19_USA Round 19 separable config (63 age bins, 4 488 FOI cells), peak GPU memory drops by ≥ 80 % for the FOI-related portion.
- [ ] No regression on existing op_system test suite.

## Files touched (estimated)

- `op_system/src/op_system/_ir.py` — new IR node
- `op_system/src/op_system/_ir_expand.py` — preserve LowRankKernel through expansion
- `op_system/src/op_system/_ir_lower.py` — codegen
- `op_system/src/op_system/_vectorize.py` — vectorized variant
- `op_system/src/op_system/compile.py` — wire into evaluator
- New tests under `op_system/tests/test_low_rank_kernel.py`

Downstream (separate PR in COVID19_USA):
- `model_input/plugins/parametric.py` — emit factors instead of dense matrix
- `configs/SMH_R19_op_system_continuum_hier_bspline_separable.yml` — switch to `type: low_rank_sum`

## Effort & risk

- **Effort**: ~5 days
- **Risk**: LOW — new primitive, dense path remains as fallback
- **Priority**: HIGH (blocks GPU prior predictive runs for Round 19 separable / joint bspline configs)

## Related

- Companion issue: "Alias inlining without CSE in vectorizer" (the two together cut FOI intermediates from 254 MB → ~1 MB per draw and give 2-3× wall-time).
